packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts

```diff
@@ -0,0 +1,118 @@
+      requiresTypeChecking: true,
+    },
+    schema: [],
+    messages: {
+      switchIsNotExhaustive:
+        'Switch is not exhaustive. Cases not matched: {{missingBranches}}',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const service = getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    function getNodeType(node: TSESTree.Node): ts.Type {
+      const tsNode = service.esTreeNodeToTSNodeMap.get(node);
+      return getConstrainedTypeAtLocation(checker, tsNode);
+    }
+
+    function fixSwitch(
+      fixer: TSESLint.RuleFixer,
+      node: TSESTree.SwitchStatement,
+      missingBranchTypes: Array<ts.Type>,
+    ): TSESLint.RuleFix | null {
+      // It seems, there is no way to insert autogenerated cases inside an empty switch body
+      if (node.cases.length === 0) {
+        return null;
+      }
+
+      const lastCase = node.cases[node.cases.length - 1];
+      const lastCaseIndent = Array(lastCase.loc.start.column + 1).join(' ');
+      let fixString = '';
+
+      for (const missingBranchType of missingBranchTypes) {
+        // While running this rule on checker.ts of TypeScript project
+        // the fix introduced a compiler error due to:
+        //
+        // type __String = (string & {
+        //         __escapedIdentifier: void;
+        //     }) | (void & {
+        //         __escapedIdentifier: void;
+        //     }) | InternalSymbolName;
+        //
+        // The following check hotfixes it.
+        if (missingBranchType.isIntersection()) {
+          continue;
+        }
+
+        const caseTest = checker.typeToString(missingBranchType);
+        const errorMessage = `Not implemented yet: ${caseTest} case`;
+
+        fixString += `\n${lastCaseIndent}case ${caseTest}: { throw new Error('${errorMessage}') }`;
+      }
+
+      return fixer.insertTextAfter(lastCase, fixString);
+    }
+
+    function checkSwitchExhaustive(node: TSESTree.SwitchStatement): void {
+      const discriminantType = getNodeType(node.discriminant);
+
+      if (discriminantType.isUnion()) {
+        const unionTypes = unionTypeParts(discriminantType);
+        const caseTypes: ts.Type[] = [];
+        for (const switchCase of node.cases) {
+          if (switchCase.test === null) {
+            // Switch has 'default' branch - do nothing.
+            return;
+          }
+
+          caseTypes.push(getNodeType(switchCase.test));
+        }
+
+        const missingBranchTypes = unionTypes.filter(
+          unionType => !caseTypes.includes(unionType),
+        );
+
+        if (missingBranchTypes.length === 0) {
+          // All cases matched - do nothing.
+          return;
+        }
+
+        context.report({
+          node: node.discriminant,
+          messageId: 'switchIsNotExhaustive',
+          data: {
+            missingBranches: missingBranchTypes
+              .map(missingType => checker.typeToString(missingType))
+              .join(', '),
```

Comment:
Might be better to instead use this, as it looks more like the union they're matching?
```suggestion
              .join(' | '),
```